compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala

```diff
@@ -0,0 +1,358 @@
+      forwarderDefSymbol0.info = NamerOps.methodType(newParamSymLists0, newResType)
+      forwarderDefSymbol0.setParamss(newParamSymLists0)
+      forwarderDefSymbol0 -> newParamSymLists0
+    }
+
+    def extractParamSymss[T](onSymbol: Symbol => T): List[List[T]] =
+      defdef.paramss.zipWithIndex.map{ case (ps, i) =>
+        if (i == annotatedParamListIndex) ps.take(paramIndex).map(p => onSymbol(p.symbol))
+        else ps.map(p => onSymbol(p.symbol))
+      }
+
+    val paramCount = defdef.symbol.paramSymss(annotatedParamListIndex).size
+
+    val (forwarderDefSymbol, newParamSymLists) = initNewForwarder()
+
+    def forwarderRhs(): tpd.Tree = {
+      val defaultOffset = defdef.paramss
+        .iterator
+        .take(annotatedParamListIndex)
+        .filter(!isTypeClause(_))
+        .map(_.size)
+        .sum
+
+      val defaultCalls = Range(paramIndex, paramCount).map(n =>
+
+        def makeSelect(refTree: Tree, name: TermName): Tree =
+          val sym = refTree.symbol
+          if !sym.findMember(name, NoPrefix, EmptyFlags, EmptyFlags).exists then
+            val param = defdef.paramss(annotatedParamListIndex)(n)
+            val methodStr = s"method ${defdef.name} ${symLocation(defdef.symbol)}"
+            val paramStr = s"parameter ${param.name}"
+            val errorMessage =
+              i"Cannot unroll $methodStr because $paramStr needs a default value"
+            report.error(errorMessage, param.srcPos)
+            ref(newErrorSymbol(sym, nme.ERROR, errorMessage.toMessage))
+          else
+            refTree.select(name)
+
+        val inner = if (defdef.symbol.isConstructor) {
+          makeSelect(ref(defdef.symbol.owner.companionModule),
+            DefaultGetterName(defdef.name, n + defaultOffset))
+        } else if (isCaseApply) {
+          makeSelect(ref(defdef.symbol.owner.companionModule),
+            DefaultGetterName(termName("<init>"), n + defaultOffset))
+        } else {
+          makeSelect(This(defdef.symbol.owner.asClass),
+            DefaultGetterName(defdef.name, n + defaultOffset))
+        }
+
+        newParamSymLists
+          .take(annotatedParamListIndex)
+          .map(_.map(ref))
+          .foldLeft(inner): (lhs, newParams) =>
+            if (newParams.headOption.exists(_.isInstanceOf[TypeTree])) TypeApply(lhs, newParams)
+            else Apply(lhs, newParams)
+      )
+
+      val forwarderInner: Tree =
+        This(defdef.symbol.owner.asClass).select(defdef.symbol)
+
+      val forwarderCallArgs =
+        newParamSymLists.zipWithIndex.map{case (ps, i) =>
+          if (i == annotatedParamListIndex) ps.map(ref).take(nextParamIndex) ++ defaultCalls
+          else ps.map(ref)
+        }
+
+      val forwarderCall0 = forwarderCallArgs.foldLeft[Tree](forwarderInner){
+        case (lhs: Tree, newParams) =>
+          if (newParams.headOption.exists(_.isInstanceOf[TypeTree])) TypeApply(lhs, newParams)
+          else Apply(lhs, newParams)
+      }
```

Comment:
The same logic can already be fond on line 197. Factor out? In fact, isn't there already an overload of `appliedTo` that handles this?

Reply:
good catch - was from copied code